blog-post/blog-post.md

```diff
@@ -0,0 +1,522 @@
+
+```
+function processPayment(Payment calldata payment) external {
+    if (payment.paymentType == PaymentType.NATIVE) {
+        require(payment.token == address(0), "Native: no token");
+        require(payment.amount > 0, "Native: amount required");
+        require(payment.tokenId == 0, "Native: no tokenId");
+        payable(payment.to).transfer(payment.amount);
+  } else if (payment.paymentType == PaymentType.ERC20) {
+        require(payment.token != address(0), "ERC20: token required");
+        require(payment.amount > 0, "ERC20: amount required");
+        require(payment.tokenId == 0, "ERC20: no tokenId");
+        IERC20(payment.token).transferFrom(payment.from, payment.to, payment.amount);
+  } else if (payment.paymentType == PaymentType.ERC721) {
+        require(payment.token != address(0), "ERC721: token required");
+        require(payment.amount == 1, "ERC721: amount must be 1");
+        require(payment.tokenId > 0, "ERC721: tokenId required");
+        IERC721(payment.token).transferFrom(payment.from, payment.to, payment.tokenId);
+  }
+}
+```
+
+Now, let's turn our attention to the encoding of this example in Core Solidity.
+The complete encoding of this example can be found [here](payment.sol).
+
+First, we start by defining an algebraic data type which represents each payment
+type by a separate data constructor with precisely the fields required for that
+specific standard. This makes invalid states **unrepresentable** by design. As an
+example, you cannot create a `Native` payment with a token field or an `ERC721`
+payment with an amount field.
+
+```
+data tokenid = tokenid(word);
+
+data Payment =
+    Native(address, word)
+  | ERC20(address, address, address, word)
+  | ERC721(address, address, address, tokenid);
+```
+
+The implementation of functions `processPayment` and `calculateFee` can be made
+much simpler by using pattern matching:
+
+```
+function processPayment(payment : Payment) {
+    match payment {
+    | Native(to,amount) =>
+        transfer(to,amount);
+    | ERC20(token,from,to,amount) =>
+        transferFromERC20(from, to, amount);
+    | ERC721(token,from,to,tokenId) =>
+        transferFromERC721(from, to, tokenId);
+    }
+}
+```
+
+Another effect of the use of pattern matching / algebraic data types is that,
+since it is not possible to represent invalid states, no runtime validations
+using `require` is necessary, since they are enforced by Core Solidity type system.
+
+In this short example, we can see how Core Solidity gives the developer the tools
+to write shorter and safer code. By using algebraic data types and pattern matching,
+we can avoid the manipulation of invalid states and the need of runtime validations
+using `require`. Also, parametric polymorphism (a.k.a. generics) and type classes
+open up new possibilities for the development of safer libraries for the modular
+development of smart-contracts.
+
+### Compile Time Evaluation
```

Comment:
This is now a subsection of `Extended example: Classic Solidity vs Core Solidity`, but does not seem connected to it. Shouldn't it be a separate one?

Reply:
We also have a small bit on compile-time evaluation under `What's next?`. I think we should merge the two.

I actually wanted to comment on that one to clarify that compile-time evaluation is an idea that we had much earlier and didn't come from Zig. What came from Zig are some specific ideas on how it could work, but those also mostly match what we were planning anyway.